### ISSUE TITLE
Precompilation significantly slows Julia startup

### ISSUE THREAD

**Issue by binarybana** [stevengj/PyCall.jl#113]
Julia startup time with PyCall precompiled into base/userimg.jl jumps in commit d19995c5fd895aa6a249cb6b6d1164dc2826e8f2 from <1 second to ~10 seconds on several of my machines (with both Julia 0.3.6 and some 0.4 masters).

I found this with a git bisect using a quick and dirty script:
```bash
#!/bin/bash
# GOOD PyCall: 896f82a
# BAD PyCall: ebe352f

JIFILE=/tmp/test_bisect
julia /home/jason/src/julia/contrib/build_sysimg.jl $JIFILE native /tmp/userimg.jl --force > /dev/null

T1=`date +%s`
/bin/time -f %U julia -J ${JIFILE}.ji -E "0"
TIME=$(echo `date +%s`-$T1 | bc -l)
echo $TIME

if [[ `echo "$TIME>2" | bc -l` -eq 1 ]]; then
        echo "Long"
        exit 1
else
        echo "Short"
        exit 0
fi
```
I'm not actually sure if this is a problem with PyCall or Julia, so if the latter, then let me know and I'll move the bug up the stack.

I don't know enough about PyCall to debug this further, but I did do some macro level debugging in [this mailing list thread](https://groups.google.com/forum/#!topic/julia-dev/uvEgky_QKmY). At the time I did not know that it was from PyCall specifically.

**Comment by stevengj** [comment]
tres bizarre...

**Comment by staticfloat** [comment]
My guess is that something weird is going on with the parser.  Is it possible for you to profile Julia and see where things are getting stuck?  I'm thinking with a C profiler, not the Julia profiler.

**Comment by binarybana** [comment]
@staticfloat , I have callgraph (with counts) output available [here](https://groups.google.com/forum/#!topic/julia-dev/uvEgky_QKmY) but let me know if you need another format.

**Comment by staticfloat** [comment]
Yeah, I honestly have no idea here.  I mean, obviously, the pycall initialization code was touched in that commit, but there's too much code for me to say what could have caused a difference in startup time.

**Comment by vtjnash** [comment]
perhaps you just need to add a line `precompile(__init__,())`, although I was pretty sure that is done automatically anyways. i'm not sure how much work that `__init__` function is trying to do, but it does look pretty long (several codegen steps are super-linear in the number of expressions)

**Comment by binarybana** [comment]
I'll try that and more once I get on the sunny side of my dissertation defense, job interviews etc... Meet you all on the other side. :)

**Comment by stevengj** [comment]
@vtjnash, the `__init__` function in PyCall doesn't actually do anything: PyCall doesn't initialize Python until the first Python routine is called, or until you manually call `pyinitialize()`.    So, you probably need `precompile(pyinitialize,(Ptr{Void},))` and probably some others.

Initially, this seemed important in order to allow you to call `pyinitialize("/path/to/python")` to specify what Python you want to use.  It seems less important now that you can just set the `PYTHON` environment variable.  However, it is still useful for the `pyjulia` Python package so that you can call `pyinitialize(C_NULL)` and tell it to use the `libpython` library that is already loaded, though I suppose I could use some environment variable to indicate that behavior.

**Comment by tomaskrehlik** [comment]
I have similar experience with homebrew build and `userimg.jl` containing `require("Lint")`. Maybe that would be easier to analyze is it is only Julia code. If I can be of any help, let me know.

**Comment by dhoegh** [comment]
I cannot reproduce this on windows on latest julia master, maybe it has gone away together with #116.

**Comment by tomaskrehlik** [comment]
Nope, `require("Lint")` still increases the time for me with the `brew` build on the OS X.
